fuzzer/src/main.rs

```diff
@@ -0,0 +1,443 @@
+                        // We assume a loop / too long behaviour in pulldown-cmark.
+                        // We print the pattern and exec ourselves again to not infinite-loop a thread.
+                        // See <https://github.com/raphlinus/pulldown-cmark/pull/286#issuecomment-490315582>
+                        // for more information.
+                        println!(
+                            "Thread timeout triggered (thread took too long) for Pattern: {:?}\n\
+                            Restarting process...",
+                            pattern,
+                        );
+                        let args: Vec<_> = env::args().collect();
+                        Command::new(&args[0])
+                            .args(&args[1..])
+                            .exec();
+                        unreachable!();
+                    }
+                }
+            }
+        }));
+
+        for thread in threads {
+            println!("Joined thread: {:?}", thread.join());
+        }
+    }).unwrap();
+}
+
+/// Function executed in worker-threads. Infinite loop generating and testing patterns.
+fn worker_thread_fn(
+    literals: Vec<Vec<u8>>, num_batches_finished: &AtomicU64, start_time: &Instant,
+    mut rng: Xoshiro256Plus, pattern_time: &Mutex<(String, Instant)>,
+) {
+    let uniform = Uniform::new(0, literals.len());
+    loop {
+        // generate `BATCH_SIZE` random patterns
+        let chunks = iter::repeat_with(|| {
+            &literals[uniform.sample(&mut rng)]
+        }).chunks(COMBINATIONS);
+        let combs = chunks.into_iter().take(BATCH_SIZE);
+
+        // test those patterns, catching panics of pulldown-cmark (and our code)
+        for comb in combs {
+            let concatenated = comb.into_iter().flatten().cloned().collect::<Vec<_>>();
+            // we mustn't panic here, because that would result in the thread being killed
+            let pattern = match String::from_utf8(concatenated) {
+                Ok(pattern) => pattern,
+                Err(err) => {
+                    println!("Couldn't convert pattern to UTF-8. Err: {:?}.", err);
+                    continue;
+                }
+            };
+            // code-block just to be double-sure that the lock is dropped to not poison it
+            { *pattern_time.lock().unwrap() = (pattern.clone(), Instant::now()); }
+            let _ = test_catch_unwind(&pattern);
+        }
+
+        // print status update and throughput
+
+        let batches_finished = num_batches_finished.fetch_add(1, Ordering::Relaxed);
+        // fetch_add returns the old value
+        let batches_finished = batches_finished + 1;
+        let patterns_finished = batches_finished * BATCH_SIZE as u64;
+        let elapsed_secs = start_time.elapsed().as_secs();
+
+        // if for some reason we are super-fast, we must avoid div-by-zero
+        if elapsed_secs > 0 {
+            println!(
+                "{:<20} throughput: {} / s",
```

Comment:
The throughput is only printed once every few minutes, so a difference of a few patters per second doesn't really matter. The zero-check is because sometimes one or two threads would end up finishing the first batch pretty much instantaneously (never figured out why). The same could happen with millis. Not panicking here is very important as any panic in a thread that's not within `test_catch_unwind` just straight up kills the whole thread. Any chance of that function panicking might result in way lower throughput when running the fuzzer on the server for several days.